File: dotnet/core/src/Workflow.cs
namespace SmooAI.SmoothOperator.Core;

/// <summary>
/// Thrown when a <see cref="Workflow{TState}"/> is misconfigured (no entry, missing node) or
/// exceeds its step cap (an unbroken cycle). Mirrors the sibling engines' <c>WorkflowError</c>.
/// </summary>
public sealed class WorkflowException : Exception
{
    public WorkflowException(string message) : base(message)
    {
    }
}

/// <summary>
/// A LangGraph-inspired typed workflow graph with conditional edges. A <see cref="Workflow{TState}"/>
/// is a state machine: <b>nodes</b> transform a typed state value and <b>edges</b> — static or
/// conditional — determine the next node. The runner starts at the entry node, applies each node
/// then follows its outgoing edge, until it reaches a terminal node (an explicit end, or a node
/// with no outgoing edge), then returns the final state. A <c>maxSteps</c> cap bounds execution.
///
/// Standalone primitive — it does not touch the agent loop. The point is the seam: a multi-step
/// orchestration (parse → guardrails → retrieve → compose → …) drops in as a graph of named nodes
/// with the routing made explicit. Mirrors the sibling engines' <c>Workflow</c>.
/// </summary>
public sealed class Workflow<TState>
{
    /// <summary>Sentinel a conditional router returns to terminate the workflow.</summary>
    public const string End = "__end__";

    private readonly Dictionary<string, Func<TState, CancellationToken, Task<TState>>> _nodes = new(StringComparer.Ordinal);
    private readonly Dictionary<string, Edge> _edges = new(StringComparer.Ordinal);
    private readonly int _maxSteps;
    private string? _entry;

    public Workflow(int maxSteps = 100)
    {
        _maxSteps = maxSteps;
    }

    /// <summary>Register an async node under <paramref name="name"/>.</summary>
    public Workflow<TState> AddNode(string name, Func<TState, CancellationToken, Task<TState>> node)
    {
        _nodes[name] = node;
        return this;
    }

    /// <summary>Register a synchronous node under <paramref name="name"/>.</summary>
    public Workflow<TState> AddNode(string name, Func<TState, TState> node) =>
        AddNode(name, (state, _) => Task.FromResult(node(state)));

    /// <summary>Add a static edge <paramref name="from"/> → <paramref name="to"/>.</summary>
    public Workflow<TState> AddEdge(string from, string to)
    {
        _edges[from] = Edge.Static(to);
        return this;
    }

    /// <summary>
    /// Add a conditional edge whose <paramref name="router"/> picks the next node at runtime,
    /// returning the target node name or <see cref="End"/> to terminate.
    /// </summary>
    public Workflow<TState> AddConditionalEdge(string from, Func<TState, string> router)
    {
        _edges[from] = Edge.Conditional(router);
        return this;
    }

    /// <summary>Set the entry node (first to execute).</summary>
    public Workflow<TState> SetEntry(string name)
    {
        _entry = name;
        return this;
    }

    /// <summary>Mark <paramref name="from"/> terminal — reaching it ends the workflow.</summary>
    public Workflow<TState> SetEnd(string from)
    {
        _edges[from] = Edge.Terminal();
        return this;
    }

    /// <summary>
    /// Execute the workflow from the entry node, returning the final state. Throws
    /// <see cref="WorkflowException"/> if no entry was set, a referenced node does not exist, or
    /// the step cap is exceeded.
    /// </summary>
    public async Task<TState> RunAsync(TState initialState, CancellationToken cancellationToken = default)
    {
        if (_entry is null)
        {
            throw new WorkflowException("workflow has no entry node — call SetEntry()");
        }
        if (!_nodes.ContainsKey(_entry))
        {
            throw new WorkflowException($"entry node '{_entry}' not found in registered nodes");
        }

        var state = initialState;
        var current = _entry;

        for (var step = 0; step < _maxSteps; step++)
        {
            if (!_nodes.TryGetValue(current, out var node))
            {
                throw new WorkflowException($"node '{current}' not found in workflow");
            }

            state = await node(state, cancellationToken).ConfigureAwait(false);

            if (!_edges.TryGetValue(current, out var edge) || edge.Kind == EdgeKind.Terminal)
            {
                // No outgoing edge, or an explicit end — terminate.
                return state;
            }
            if (edge.Kind == EdgeKind.Conditional)
            {
                var target = edge.Router!(state);
                if (target == End)
                {
                    return state;
                }
                current = target;
            }
            else
            {
                current = edge.To!;
            }
        }

        throw new WorkflowException($"workflow exceeded maxSteps ({_maxSteps}) — possible infinite loop");
    }

    private enum EdgeKind
    {
        Static,
        Conditional,
        Terminal,
    }

    private sealed record Edge(EdgeKind Kind, string? To, Func<TState, string>? Router)
    {
        public static Edge Static(string to) => new(EdgeKind.Static, to, null);

        public static Edge Conditional(Func<TState, string> router) => new(EdgeKind.Conditional, null, router);

        public static Edge Terminal() => new(EdgeKind.Terminal, null, null);
    }
}

File: python/core/src/smooth_operator_core/workflow.py
"""LangGraph-inspired typed workflow graph with conditional edges.

Phase-3 sibling of the reference engine's workflow primitive. A :class:`Workflow`
is a state machine: **nodes** transform a typed state value and **edges** — static
or **conditional** — determine the next node to execute. The runner starts at the
entry node, applies each node then follows its outgoing edge, until it reaches an
``END`` sentinel (or a node with no outgoing edge), then returns the final state.

Nodes may be sync or async; the runner awaits coroutine results. A ``max_steps``
cap bounds execution so an intentional or accidental cycle can't loop forever.

This is a standalone module — it does not touch the agent loop. The point is the
seam: a multi-step orchestration (parse → guardrails → retrieve → compose → …)
drops in as a graph of named nodes with the routing made explicit.
"""

from __future__ import annotations

import inspect
from collections.abc import Awaitable
from typing import Callable, Generic, TypeVar, Union

S = TypeVar("S")

# Sentinel a conditional router can return to signal termination.
END: str = "__end__"

# A node transforms state into a new state; it may return a value or a coroutine.
NodeFn = Callable[[S], Union[S, Awaitable[S]]]

# A conditional router inspects the current state and returns the next node name
# (or ``END`` to terminate).
Router = Callable[[S], str]


class WorkflowError(Exception):
    """Raised when a workflow is misconfigured or exceeds its step limit."""


class Workflow(Generic[S]):
    """A typed workflow graph: named nodes connected by static/conditional edges.

    Build with :meth:`add_node`, :meth:`add_edge` / :meth:`add_conditional_edge`,
    :meth:`set_entry`, and :meth:`set_end`; the builder methods return ``self`` so
    they chain. :meth:`run` executes the graph from the entry node.
    """

    def __init__(self, max_steps: int = 100) -> None:
        self._nodes: dict[str, NodeFn[S]] = {}
        # An edge is either a node name (static) or a Router (conditional). ``END``
        # marks a terminal node.
        self._edges: dict[str, Union[str, Router[S]]] = {}
        self._entry: str | None = None
        self._max_steps = max_steps

    def add_node(self, name: str, func: NodeFn[S]) -> Workflow[S]:
        """Register a node ``func`` under ``name`` (used to reference it in edges)."""
        self._nodes[name] = func
        return self

    def add_edge(self, from_node: str, to_node: str) -> Workflow[S]:
        """Add a static edge ``from_node`` → ``to_node``."""
        self._edges[from_node] = to_node
        return self

    def add_conditional_edge(self, from_node: str, router: Router[S]) -> Workflow[S]:
        """Add a conditional edge whose ``router`` picks the next node at runtime.

        The router receives the current state and returns the target node name, or
        ``END`` to terminate the workflow.
        """
        self._edges[from_node] = router
        return self

    def set_entry(self, name: str) -> Workflow[S]:
        """Set the entry node (first node to execute)."""
        self._entry = name
        return self

    def set_end(self, from_node: str) -> Workflow[S]:
        """Mark ``from_node`` as terminal — reaching it ends the workflow."""
        self._edges[from_node] = END
        return self

    async def run(self, initial_state: S) -> S:
        """Execute the workflow from the entry node, returning the final state.

        Raises :class:`WorkflowError` if no entry node was set, a referenced node
        does not exist, or the ``max_steps`` cap is exceeded (e.g. an unbroken
        cycle).
        """
        if self._entry is None:
            raise WorkflowError("workflow has no entry node — call set_entry()")
        if self._entry not in self._nodes:
            raise WorkflowError(f"entry node '{self._entry}' not found in registered nodes")

        state = initial_state
        current = self._entry

        for _ in range(self._max_steps):
            node = self._nodes.get(current)
            if node is None:
                raise WorkflowError(f"node '{current}' not found in workflow")

            result = node(state)
            state = await result if inspect.isawaitable(result) else result

            edge = self._edges.get(current)
            if edge is None:
                # No outgoing edge — implicit end.
                return state
            if edge == END:
                return state
            if callable(edge):
                target = edge(state)
                if target == END:
                    return state
                current = target
            else:
                current = edge

        raise WorkflowError(f"workflow exceeded max_steps ({self._max_steps}) — possible infinite loop")
